### PR TITLE
KAFKA-7050: Decrease default consumer request timeout to 30s

### PR DESCRIPTION
**Pull request by hachikuji**
This patch changes the default `request.timeout.ms` of the consumer to 30 seconds. Additionally, it adds logic to `NetworkClient` to set timeouts at the request level. We use this to handle the special case of the JoinGroup request, which may block for as long as the value configured by `max.poll.interval.ms`.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/clients/NetworkClient.java [hunk 1]
```diff
@@ -89,7 +89,7 @@
     private int correlation;
 
     /* max time in ms for the producer to wait for acknowledgement from server*/
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/NetworkClient.java]: This seems outdated.

File: clients/src/main/java/org/apache/kafka/clients/InFlightRequests.java [hunk 2]
```diff
@@ -178,7 +177,7 @@ public boolean isEmpty() {
             if (!deque.isEmpty()) {
                 NetworkClient.InFlightRequest request = deque.peekLast();
                 long timeSinceSend = now - request.sendTimeMs;
-                if (timeSinceSend > requestTimeoutMs)
+                if (timeSinceSend > request.requestTimeoutMs)
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/InFlightRequests.java]: Can you explain the reasoning of just looking at the last item in the queue? It may be worth adding a comment.

**Review by ijuma — COMMENTED** [review]
Thanks for the PR, a couple of comments.

File: clients/src/main/java/org/apache/kafka/clients/NetworkClient.java [hunk 3]
```diff
@@ -89,7 +89,7 @@
...
-    private final int requestTimeoutMs;
+    private final int minRequestTimeoutMs;
```
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/NetworkClient.java]: nit: update the comment above is to be more consistent with the name

File: clients/src/main/java/org/apache/kafka/clients/NetworkClient.java [hunk 4]
```diff
@@ -117,14 +117,14 @@ public NetworkClient(Selectable selector,
...
-                         int requestTimeoutMs,
+                         int minRequestTimeoutMs,
                          Time time,
                          boolean discoverBrokerVersions,
                          ApiVersions apiVersions,
                          LogContext logContext) {
         this(null, metadata, selector, clientId, maxInFlightRequestsPerConnection,
              reconnectBackoffMs, reconnectBackoffMax,
-             socketSendBuffer, socketReceiveBuffer, requestTimeoutMs, time,
+             socketSendBuffer, socketReceiveBuffer, minRequestTimeoutMs, time,
```
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/NetworkClient.java]: Use one line per parameter? Ditto below.

File: clients/src/main/java/org/apache/kafka/clients/NetworkClient.java [hunk 5]
```diff
@@ -1015,8 +1022,28 @@ public ClientRequest newClientRequest(String nodeId, AbstractRequest.Builder<?>
...
+        final long requestTimeoutMs;
+
+        public InFlightRequest(ClientRequest clientRequest,
+                               RequestHeader header,
+                               boolean isInternalRequest,
+                               AbstractRequest request,
+                               Send send,
+                               long sendTimeMs) {
+            this(header,
+                    clientRequest.requestTimeoutMs(),
```
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/NetworkClient.java]: nit: misalignment.

File: clients/src/main/java/org/apache/kafka/clients/NetworkClient.java [hunk 6]
```diff
@@ -475,7 +472,7 @@ private void doSend(ClientRequest clientRequest, boolean isInternalRequest, long
 
         long metadataTimeout = metadataUpdater.maybeUpdate(now);
         try {
-            this.selector.poll(Utils.min(timeout, metadataTimeout, requestTimeoutMs));
+            this.selector.poll(Utils.min(timeout, metadataTimeout, minRequestTimeoutMs));
```
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/NetworkClient.java]: I'm wondering if it still makes sense to consider metadata timeout and request timeout to determine the timeout for `selector.poll`, but maybe it can be discussed in another PR..

File: clients/src/test/java/org/apache/kafka/clients/MockClient.java [hunk 7]
```diff
@@ -296,6 +297,15 @@ private synchronized void maybeAwaitWakeup() {
         return copy;
     }
 
+    private void checkTimeoutOfPendingRequests(long nowMs) {
+        ClientRequest request = requests.peek();
+        while (request != null && nowMs - request.createdTimeMs() > request.requestTimeoutMs()) {
```
Comment (guozhangwang) [inline clients/src/test/java/org/apache/kafka/clients/MockClient.java]: Take care of overflow with `remainingTimeAtLeastZero`

File: clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java [hunk 8]
```diff
@@ -1754,9 +1756,11 @@ private void makeCoordinatorUnknown(ConsumerCoordinator coordinator, Errors erro
         consumerClient.poll(0);
         assertTrue(coordinator.coordinatorUnknown());
     }
+
     private void closeVerifyTimeout(final ConsumerCoordinator coordinator,
-            final long closeTimeoutMs, final long requestTimeoutMs,
-            long expectedMinTimeMs, long expectedMaxTimeMs) throws Exception {
+                                    final long closeTimeoutMs,
+                                    long expectedMinTimeMs,
```
Comment (guozhangwang) [inline clients/src/test/java/org/apache/kafka/clients/consumer/internals/ConsumerCoordinatorTest.java]: nit: add final for other two parameter?

File: core/src/main/scala/kafka/common/InterBrokerSendThread.scala [hunk 9]
```diff
@@ -118,9 +119,9 @@ abstract class InterBrokerSendThread(name: String,
 
   private def failExpiredRequests(now: Long): Unit = {
     // clear all expired unsent requests
-    val expiredRequests = unsentRequests.removeExpiredRequests(now, unsentExpiryMs)
+    val expiredRequests = unsentRequests.removeExpiredRequests(now, requestTimeoutMs)
```
Comment (guozhangwang) [inline core/src/main/scala/kafka/common/InterBrokerSendThread.scala]: nit: update `def removeExpiredRequests(now: Long, unsentExpiryMs: Long)` to `removeTimedoutRequest(.., requestTimouetMs: Long)`?

**Review by guozhangwang — COMMENTED** [review]
Could we trigger 5 times of streams system test to double check on any resilience regressions?

File: clients/src/main/java/org/apache/kafka/clients/InFlightRequests.java [hunk 10]
```diff
@@ -169,7 +169,7 @@ public boolean isEmpty() {
      * @return list of nodes
      */
     public List<String> getNodesWithTimedOutRequests(long now) {
-        List<String> nodeIds = new LinkedList<>();
+        List<String> nodeIds = new ArrayList<>();
```
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/InFlightRequests.java]: Hah, I had done this too. Any opportunity to remove `LinkedList`, eh? ;)

**Comment by hachikuji** [comment]
@ijuma @guozhangwang I think this is ready for another look.

**Comment by guozhangwang** [comment]
LGTM!

**Comment by hachikuji** [comment]
@guozhangwang I ran through streams system tests a few times and saw no failures.

**Comment by guozhangwang** [comment]
> @guozhangwang I ran through streams system tests a few times and saw no failures.

Thank you!

**Review by ijuma — APPROVED** [review]
Since @guozhangwang has already reviewed, I briefly looked only. As discussed offline, the semantics are a bit weird (i.e. some requests could be cancelled quite quickly depending on other requests), but since we have to handle disconnections anyway, maybe this is fine for now. We can revisit later if other clients with different requirements are updated to use this (like the AdminClient).

LGTM.

**Comment by hachikuji** [comment]
Thanks @ijuma and @guozhangwang. On the point about timeout semantics, I think this is not so much a consequence of using different timeouts, but of allowing multiple in-flight requests on the same socket. Even if the timeouts are all identical, one request can arrive just as a current in-flight is about to be expired. What we should try to avoid is having client behavior dependent on these timeouts, which means decoupling the request timeout from application timeouts and retrying when needed.

**Comment by ijuma** [comment]
Yeah, I agree that the complexity comes from the having multiple requests in flight in the same socket. I also agree that the application timeouts and retries may be the way forward. As I said, we need it for socket disconnections, so it seems reasonable to take advantage of it for this case too.

Finally, the alternative option where we never timeout a request before it has truly expired seems to be too problematic as it could cause requests to never time out given a steady stream of requests.
